src/cart/crt_hg.c

```diff
@@ -589,6 +811,133 @@ crt_hg_init(crt_phy_addr_t *addr, bool server)
 	return rc;
 }
 
+int
+crt_dump_ha_list(d_list_t *halist)
+{
+	struct crt_ha_list_entry	*entry;
+	struct crt_ha_list_entry	*temp;
+	struct ha_entry			ha_value;
+	int				rc = 0;
+
+	d_list_for_each_entry_safe(entry, temp, halist,
+					chl_list_link) {
+		ha_value = entry->chl_entry;
+
+		D_DEBUG(DB_TRACE, "hg addr %" PRIu64 "\n",
+				(uint64_t)ha_value.ha_addr);
+		D_DEBUG(DB_TRACE, "hg class %" PRIu64 "\n\n",
+				(uint64_t) ha_value.ha_class);
+	}
+	return rc;
+}
+
+
+int
+crt_hg_ha_list_empty(d_list_t *halink, void *arg)
+{
+	struct crt_ha_mapping		*ha;
+	struct crt_ha_list_entry	*entry;
+	struct crt_ha_list_entry	*temp;
+	struct ha_entry			ha_value;
+	int				rc = 0;
+
+	D_ASSERT(arg == NULL);
+	ha = crt_ha_link2ptr(halink);
+
+	if (ha->chm_initialized != 1)
+		D_GOTO(out, rc);
+
+	D_DEBUG(DB_TRACE, "client id %" PRIu64 " removing from hash table\n",
+				ha->chm_key);
+
+	d_list_for_each_entry_safe(entry, temp, &ha->chm_list,
+					chl_list_link) {
+		ha_value = entry->chl_entry;
+		rc = crt_hg_remove_addr(ha_value.ha_class, ha_value.ha_addr);
+
+		if (rc != HG_SUCCESS) {
+			D_ERROR("crt_remove_addr failed, hg_ret %d.\n", rc);
+		}
+		D_FREE(entry);
+	}
+out:
+	return rc;
+}
+
+int
+crt_hg_remove_hash_rank(d_rank_t rank)
+{
+	d_list_t			*halink;
+	struct crt_ha_mapping		*ha;
+	struct ha_entry			ha_value;
+	struct crt_ha_list_entry	*entry;
+	uint64_t			client_id;
+	int				rc = 0;
+	struct crt_context		*crt_ctx;
+
+/*	if (!crt_is_service())
+*		D_GOTO(out, rc);
+*/
+	if (crt_gdata.cg_na_plugin != CRT_NA_OFI_PSM2)
+		D_GOTO(out, rc);
+	D_RWLOCK_RDLOCK(&crt_gdata.cg_rwlock);
+
+	client_id = (uint64_t)rank;
+	d_list_for_each_entry(crt_ctx, &crt_gdata.cg_ctx_list, cc_link) {
+		/*Can this lock create a deadlock? TBD*/
+		D_RWLOCK_WRLOCK(&crt_ctx->cc_ha_server_hash_table_rwlock);
+		halink = d_hash_rec_find(&crt_ctx->cc_ha_server_hash_table,
+				(void *)&client_id, sizeof(client_id));
+		if (!halink) {
```

Comment:
is this an error though? we can have entry in one context table but not in others. if this doesnt find in first table it will skip all other context tables.

The scenario where this can happen:
servers have multiple contexts
server0 sends rpc to server1 on tag=3
this will result in server1's context[3]  table to be populated with server0's address/info, but no other context table will be

Reply:
i will just log the error and do a continue;